packages/spotlight/src/server/utils/cors.ts

```diff
@@ -1,15 +1,273 @@
+ *
+ * Results are cached for 1 minute to handle network changes while
+ * avoiding repeated system calls.
+ */
+function getMachineIPs(): Set<string> {
+  const now = Date.now();
+  if (machineIPsCache && now - machineIPsCacheTime < MACHINE_IPS_CACHE_TTL_MS) {
+    return machineIPsCache;
+  }
+
+  const ips = new Set<string>();
+  const interfaces = os.networkInterfaces();
+
+  for (const name in interfaces) {
+    const addresses = interfaces[name];
+    if (addresses) {
+      for (const addr of addresses) {
+        // Add the IP address (normalized)
+        ips.add(addr.address);
+      }
+    }
+  }
+
+  machineIPsCache = ips;
+  machineIPsCacheTime = now;
+  return ips;
+}
+
+/**
+ * Check if a string is an IP address (IPv4 or IPv6).
+ * Uses Node's net module for reliable detection.
+ */
+function isIPAddress(hostname: string): boolean {
+  // Handle bracketed IPv6 (e.g., [::1])
+  const cleanHostname = hostname.startsWith("[") && hostname.endsWith("]") ? hostname.slice(1, -1) : hostname;
+  return net.isIP(cleanHostname) !== 0;
+}
+
+/**
+ * Check if an IP address belongs to this machine.
+ *
+ * This checks against IPs assigned to the machine's network interfaces,
+ * which includes:
+ * - Loopback (127.0.0.1, ::1)
+ * - Local network IPs (192.168.x.x, 10.x.x.x, etc.)
+ * - VPN IPs (e.g., Tailscale 100.x.x.x, Zerotier IPs)
+ *
+ * This allows services like custom domains pointing to the machine,
+ * or VPN-based access, to work with Spotlight.
+ *
+ * Security note: This is safe because:
+ * - These are IPs the machine actually has assigned
+ * - External access requires explicit port forwarding/exposure
+ * - The 6-hour DNS cache still protects against rebinding attacks
+ */
+function isLocalMachineIP(ip: string): boolean {
+  // Handle bracketed IPv6
+  const cleanIP = ip.startsWith("[") && ip.endsWith("]") ? ip.slice(1, -1) : ip;
+
+  // Check if it's one of the machine's own IPs
+  const machineIPs = getMachineIPs();
+  return machineIPs.has(cleanIP);
+}
+
+/**
+ * Resolve a hostname to IP addresses using both DNS and OS resolver.
+ *
+ * Why both dns.resolve() and dns.lookup()?
+ * - dns.resolve4/6: Direct DNS query, returns TTL when available
+ * - dns.lookup: Uses OS resolver (getaddrinfo), which reads /etc/hosts
+ *
+ * This ensures we handle both:
+ * - Standard DNS hostnames
+ * - Local hostnames defined in /etc/hosts (common in development)
+ *
+ * @returns Array of resolved IPs with optional TTL, and the minimum TTL found
+ */
+async function resolveHostname(hostname: string): Promise<{ addresses: DnsResult[]; minTtl?: number }> {
+  const results: DnsResult[] = [];
+  let minTtl: number | undefined;
+
+  // Run all resolution methods in parallel for efficiency
+  const [resolve4Result, resolve6Result, lookupResult] = await Promise.allSettled([
+    resolve4(hostname, { ttl: true }),
+    resolve6(hostname, { ttl: true }),
+    lookup(hostname, { all: true }),
+  ]);
+
+  // Process IPv4 DNS results (with TTL)
```

Comment:
The `Math.max()` along with `minTtl` logic in the entire file seems to be incorrect. We want to _reject_ DNS records with a TTL < MIN_TTL_MS as those records are unsafe. Forcing our own longer TTL sounds actually wrong.

That said we should probably make `MIN_TTL_MS` 1 hour